khiops/utils/dataset.py

```
@@ -29,12 +32,268 @@
+            raise ValueError(f"A relation must be of size 2 or 3, not {len(relation)}")
+
+        # Check the types of the tuple contents
+        parent_table, child_table = relation[:2]
+        if not isinstance(parent_table, str):
+            raise TypeError(
+                type_error_message(f"Relation #{i}'s parent table", parent_table, str)
+            )
+        if not isinstance(child_table, str):
+            raise TypeError(
+                type_error_message(f"Relation #{i}'s child table", child_table, str)
+            )
+        if len(relation) == 3 and not isinstance(relation[2], bool):
+            raise TypeError(
+                type_error_message(
+                    f"Relation #{i} ({parent_table}, {child_table}) 1-1 flag",
+                    relation[2],
+                    bool,
+                )
+            )
+
+    # Check structure and coherence with the rest of the spec
+    parents_and_children = [relation[:2] for relation in relations_spec]
+    for i, relation in enumerate(relations_spec, 1):
+        parent_table, child_table = relation[:2]
+        if parent_table == child_table:
+            raise ValueError(
+                f"Relation #{i}'s tables are equal: ({parent_table}, {child_table}). "
+                "They must be different."
+            )
+        for table in (parent_table, child_table):
+            if not table in tables_spec.keys():
+                raise ValueError(
+                    f"Relation #{i} ({parent_table}, {child_table}) contains "
+                    f"non-existent table '{table}'. All relation tables must exist "
+                    "in the 'tables' entry."
+                )
+        if parents_and_children.count(relation[:2]) > 1:
+            raise ValueError(
+                f"Relation #{i} ({parent_table}, {child_table}) occurs "
+                f"{parents_and_children.count(relation[:2])} times. "
+                f"Each relation must be unique."
+            )
```

I would put this `if` before the previous `for`, thus the check could fail faster, without needing to go through the loop over all the secondary tables.

---

We need the `relation` iteration variable.

---

I have proposed to move this `if`, before the _inner_ `for` just above it, not before the outer for.

---

Ok, I get it but we do not need to optimize this.